### PR TITLE
FEEvaluation: make active_fe_index and active_quad_index an optional …

### PR DESCRIPTION
**Pull request by peterrum**
…input argument

This PR allows me to write:
```cpp
  {
    // active_fe_index = 0
    std::cout << FEEvaluation<dim, fe_degree, fe_degree + 1, n_components, Number, VectorizedArrayType>(matrix_free, 0, 0, 0).dofs_per_cell     << std::endl;
    std::cout << FEEvaluation<dim, -1, 0, n_components, Number, VectorizedArrayType>(matrix_free, 0, 0, 0, 0, 0).dofs_per_cell                  << std::endl;

    // active_fe_index = 1
    std::cout << FEEvaluation<dim, fe_degree + 1, fe_degree + 1, n_components, Number, VectorizedArrayType>(matrix_free, 0, 0, 0).dofs_per_cell << std::endl;
    std::cout << FEEvaluation<dim, -1, 0, n_components, Number, VectorizedArrayType>(matrix_free, 0, 0, 0, 1, 0).dofs_per_cell                  << std::endl;
  }
```

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -3250,36 +3267,48 @@ inline FEEvaluationBaseData<dim, Number, is_face, VectorizedArrayType>::
...
+               active_quad_index)
+             .n_q_points_face :
+           data_in
+             .get_shape_info(
+               dof_no,
+               quad_no_in,
+               dof_info->component_to_base_index[first_selected_component],
+               active_fe_index,
+               active_quad_index)
+             .n_q_points))
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_evaluation.h:3311]: part of #11213.

File: tests/matrix_free/matrix_vector_hp_no_template.cc [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+      {
+        const auto cell_subrange =
+          data.create_cell_subrange_hp_by_index(cell_range, i);
+
+        if (cell_subrange.second <= cell_subrange.first)
+          continue;
+
+        helmholtz_operator_no_template(data, dst, src, cell_subrange, i, i);
+      }
+  }
```
Comment (peterrum) [inline tests/matrix_free/matrix_vector_hp_no_template.cc:57]: ... here ...

File: tests/matrix_free/matrix_vector_mf.h [hunk 3]
```diff
@@ -56,6 +56,38 @@ helmholtz_operator(const MatrixFree<dim, typename VectorType::value_type> &data,
...
+      fe_eval.evaluate(EvaluationFlags::values | EvaluationFlags::gradients);
+      for (unsigned int q = 0; q < n_q_points; ++q)
+        {
+          fe_eval.submit_value(Number(10) * fe_eval.get_value(q), q);
+          fe_eval.submit_gradient(fe_eval.get_gradient(q), q);
+        }
+      fe_eval.integrate(EvaluationFlags::values | EvaluationFlags::gradients);
+      fe_eval.distribute_local_to_global(dst);
+    }
+}
```
Comment (peterrum) [inline tests/matrix_free/matrix_vector_mf.h:87]: ... and here.

**Review by peterrum — COMMENTED** [review]
@kronbichler What do you think about the test ...

**Review by kronbichler — APPROVED** [review]
Looks good to me, thank you for the cleanup. But I must say that we have a large number of optional arguments now that makes it a pain to recall what say the fifth integer parameter is and how it differs from the fourth. Not sure what would be a more intuitive interface, though: One could think of a fancy `UpdateFlags` kind of structure that allows to pass names and values, but I don't see yet a variant that would be much better.

**Comment by peterrum** [comment]
Travis seem to be stuck. I have already pushed a second time.

**Comment by kronbichler** [comment]
Since we run a lot of matrix-free tests in the parallel CI build (probably not fewer than in the serial case), I will not bother waiting for the serial CI and merge. This looks safe.

**Pull request merged**